### PR TITLE
Refactor reconciler to further cater to non-ARM resources

### PR DESCRIPTION
**Pull request by matthchr**
This is moving us in a positive direction but doesn't get us all the way there.
There will be further changes needed, but rather than build them all up and get it perfect, 
I'm instead sending pieces so that it's merged faster and also is somewhat more digestible.

### CODE REVIEW

File: v2/internal/reconcilers/arm/azure_generic_arm_reconciler.go [hunk 1]
```diff
@@ -46,6 +39,7 @@ const (
 	DeleteActionMonitorDelete = DeleteAction("MonitorDelete")
 )
 
+// TODO: Stopped here... noodling about if this should be a generic interface?
```
Comment (matthchr) [inline v2/internal/reconcilers/arm/azure_generic_arm_reconciler.go:42]: Remove this

File: v2/pkg/genruntime/resolved_secrets.go [hunk 2]
```diff
@@ -30,3 +30,14 @@ func (r ResolvedSecrets) LookupSecret(ref SecretReference) (string, error) {
 	}
 	return result, nil
 }
+
+// MustLookupSecret looks up the secret value for the given reference. If the reference is nil, an error is returned.
+// If the secret cannot be found, an error is returned
+// TODO: Wondering if there's a better name for this than what I've got?
```
Comment (matthchr) [inline v2/pkg/genruntime/resolved_secrets.go:36]: Remove this note... but also if anybody has a better name feel free to suggest it

Comment (theunrepentantgeek) [reply v2/pkg/genruntime/resolved_secrets.go:36]: Why is this not just `LookupSecret()`?

I thought the Go ecosystem convention for the prefix `Must` was that such methods would panic, and not return errors.

Given that this one _can_ fail, maybe `TryLookupSecret()`?

**Review by theunrepentantgeek — APPROVED** [review]
Looks good - I think I see where you're heading. Might be an idea to create some class diagrams to show the relationships between interfaces.

File: v2/internal/reconcilers/common.go [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+			"resourceVersion", obj.GetResourceVersion(),
+			"generation", obj.GetGeneration(),
+			"uid", obj.GetUID(),
+			"ownerReferences", obj.GetOwnerReferences(),
+			"creationTimestamp", obj.GetCreationTimestamp(),
+			"finalizers", obj.GetFinalizers(),
+			"annotations", ourAnnotations,
+			// Use fmt here to ensure the output uses the String() method, which log.Info doesn't seem to do by default
+			"conditions", fmt.Sprintf("%s", obj.GetConditions()),
+		}
```
[the remark this answers is not in the document]
Comment (matthchr) [reply v2/internal/reconcilers/common.go:52]: The problem is that the actual `log` call below which uses this value has the signature:
`func (l Logger) Info(msg string, keysAndValues ...interface{})`

So if I make this `map[string]interface{}`, then I can't just pass it in to the log function (which just expects a slice of interfaces).

We're no worse off here than we were before though. Prior to refactoring we called `Info` directly which lets you do whatever you want.

File: v2/pkg/genruntime/resolved_secrets.go (hunk 2 above)
Comment (matthchr) [reply v2/pkg/genruntime/resolved_secrets.go:36]: Yes you're right, `Must` is not the right prefix. The reason I'm not using `LookupSecret` is that already exists. The difference is that this ones signature accepts a ptr `*SecretReference`.

Comment (matthchr) [reply v2/pkg/genruntime/resolved_secrets.go:36]: Maybe just `LookupSecretPtr` (even though it's a bit horrible it's clearer than `Must`)

Comment (matthchr) [reply v2/pkg/genruntime/resolved_secrets.go:36]: Have gone with `LookupSecretPtr` - let me know if you think there's a better name.

**Comment by matthchr** [comment]
> Looks good - I think I see where you're heading. Might be an idea to create some class diagrams to show the relationships between interfaces.

Have filed #2295 this for tracking the relationships. They're still in flux a bit so for now I think it makes sense to evolve them in code and then once we've gotten them solidified we can draw something up.

File: v2/pkg/genruntime/resolved_secrets.go (hunk 2 above)
Comment (theunrepentantgeek) [reply v2/pkg/genruntime/resolved_secrets.go]: Ah, my kingdom for some method overloads. 😁

Comment (matthchr) [reply v2/pkg/genruntime/resolved_secrets.go]: Will do `LookupSecretFromPtr` after discussion for extra clarity.

File: v2/internal/reconcilers/common.go [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+	}
+}
+
+type ARMOwnedResourceReconcilerCommon struct {
+	ReconcilerCommon
+	ResourceResolver *resolver.Resolver
+}
+
+// IsOwnerReady returns true if the owner is ready or if there is no owner required
+func (r *ARMOwnedResourceReconcilerCommon) IsOwnerReady(ctx context.Context, log logr.Logger, obj genruntime.ARMOwnedMetaObject) (bool, error) {
```
[the remark this answers is not in the document]
Comment (matthchr) [reply v2/internal/reconcilers/common.go:71]: Done

**Pull request merged**